torchao/prototype/autoround/core.py

```diff
@@ -0,0 +1,391 @@
+        # convert [A, MultiTensor(b1,b2,b3), MultiTensor(c1,c2,c3)] => [[A,b1,c1], [A,b2,c2] [A,b3,c3]]
+        grouped_args = cls.flat_to_grouped(flat_args)
+        # run function for each of the multitensors and return a multitensor
+        outputs = []
+        with torch._C.DisableTorchFunctionSubclass():
+            # Note: for the decoder, apply auto-round to it
+            if func is torch.ops.transformers_ops.general_decoder:
+                outputs = optimize_decoder(func, grouped_args, spec)
+            else:
+                for i, inp in enumerate(grouped_args):
+                    cur_args, cur_kwargs = tree_unflatten(inp, spec)
+                    cur_args = ar_utils.move_data_to_device(cur_args, "cuda")
+                    cur_kwargs = ar_utils.move_data_to_device(cur_kwargs, "cuda")
+                    out = func(*cur_args, **cur_kwargs)
+                    outputs.append(out.cpu() if isinstance(out, torch.Tensor) else out)
+            grouped_outputs = [tree_flatten(x)[0] for x in outputs]
+            out_spec = tree_flatten(outputs[0])[1]
+            # convert [[A,b1,c1], [A,b2,c2] [A,b3,c3]] => [A, MultiTensor(b1,b2,b3), MultiTensor(c1,c2,c3)]
+            flat_outputs, non_tensors_equal = cls.grouped_to_flat(grouped_outputs)
+            assert non_tensors_equal, (
+                f"ERR: found a function in model: {func} which "
+                + "caused an error in MultiTensor, the function dispatch only works for functions"
+                + " with Tensor outputs or that have the same non-Tensor output value for all across all inputs"
+            )
+            return tree_unflatten(flat_outputs, out_spec)
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs={}):
+        pass
+
+    def __tensor_flatten__(self):
+        return ["values"], None
+
+    @classmethod
+    def __tensor_unflatten__(
+        cls, tensor_data_dict, tensor_attributes, outer_size, outer_stride
+    ):
+        cls(tensor_data_dict["values"])
+
+
+@dataclass
+class AutoRoundConfig:
+    bits: int = 4
+    sym: bool = False
+    iters: int = 200
+    group_size: int = 128
+    train_bs: int = 8
+    eval_bs: int = 4
+    seed: int = 42
+    amp: bool = False
+    nsamples: int = 128
+    seqlen: int = 2048
+
+
+auto_round_config = AutoRoundConfig()
+
+
+@torch.no_grad()
+def create_qmodel_from_qdq_model(qdq_model: torch.nn.Module):
+    """Create a quantized model from a qdq model.
+    The qdq_model includes Linear quantized by auto-round, which includes qdq weight, scale, zp.
+    """
+
+    @torch.no_grad()
+    def apply_static_quant(observed_linear):
+        device = observed_linear.weight.device
+        weight_scale = observed_linear.scale.to(device)
+        weight_zero_point = observed_linear.zp.to(device)
+
+        def weight_quant_func(weight):
+            block_size = (1, observed_linear.group_size)
+            # TODO: shift the zero and prepack the weight to use tinygemm?
```

Comment:
yeah you can try that I think, but ideally your scale/zero_point can be compatible with tinygemm in the very beginning (use preserve_zero=False and zero_point_domain to be FLOAT), otherwise there will be some accuracy loss

Reply:
Thanks, will try to use it when I refactor the `auto-round` side.